[MRG+1] fix logistic regression class weights

TomDLT opened the pull request:
-This PR tests that these two classifiers are equivalent:

``` python
class_weights = compute_class_weight("balanced", np.unique(y), y)
clf1 = LogisticRegression(multi_class="multinomial", class_weight="balanced")
clf2 = LogisticRegression(multi_class="multinomial", class_weight=class_weights)
```

-[EDIT]This PR also tests that these two classifiers are equivalent: (fix #5450)

``` python
class_weights = compute_class_weight("balanced", np.unique(y), y)
clf1 = LogisticRegressionCV(class_weight="balanced")
clf2 = LogisticRegressionCV(class_weight=class_weights)
```

That is why we can remove [this line](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/tests/test_common.py#L116)

I also : 
- updated some remaining `"auto"` (deprecated) into `"balanced"`.
- added the test from #5420, which tests the bug detailed in #5415.
- changed the variable names to distinguish `y_bin` and `Y_bin`

TomDLT committed:
FIX class_weight in LogisticRegression and LogisticRegressionCV

ogrisel commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -540,6 +540,50 @@ def test_logistic_regressioncv_class_weights():
+        clf2.fit(X, y)
+        assert_array_almost_equal(clf1.coef_, clf2.coef_, decimal=6)
+
+    # Binary case: remove 90% of class 0 and 100% of class 2
```
Why not also test multiclass OvR with 3 imbalanced classes?

ogrisel replied:
I think I understand the equivalent `class_weight_dict` for 3 classes with OvR is no longer trivial to compute in that case.

TomDLT replied:
exactly, at least we cannot used directly `compute_class_weight`

ogrisel commented:
The way to deal with `class_weight='balanced'` and `multiclass='ovr'` here is not consistent with other estimators such as `SGDClassifier` for instance.

For `SGDClassifier` the expanded `class_weight` is computed on the total classification problem by calling `compute_class_weight` once with the full multiclass `y` and then for each binary subproblem `i` is fit with `expanded_class_weight[i]` for the positive class and `1.` for the negative class (that is the "rest").

I don't know if this is principled or not but we should enforce consistency throughout the OvR classifiers.

@hannawallach @amueller any comment on this issue?

TomDLT commented:
>For `SGDClassifier` the expanded `class_weight` is computed on the total classification problem by calling `compute_class_weight` once with the full multiclass `y` and then for each binary subproblem `i` is fit with `expanded_class_weight[i]` for the positive class and `1.` for the negative class (that is the "rest").

I don't know which way is preferable, but `compute_class_weight` is used only in SGD, LogisticRegression, and some SVC classes.
I didn't get into every details of all SVC classes, but LinearSVC seems to behave as SGD, as you can see in [this line](https://github.com/TomDLT/scikit-learn/blob/logistic_multiclass/sklearn/svm/src/liblinear/linear.cpp#L2469.).

We could change LogisticRegression to behave in the same way, for consistency.
However, I don't know how to deprecate the change in a clean way.

___
This remark being said, this PR solves a small bug about `class_weight="balanced"` in the multinomial case and is not correlated to this consistency issue.

amueller commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -542,12 +542,12 @@ def test_logistic_regressioncv_class_weights():
-    # Test the liblinear fails when class_weight of type dict is
-    # provided, when it is multiclass. However it can handle
-    # binary problems.
+    msg = ("In LogisticRegressionCV the liblinear solver cannot handle "
```
Sorry I don't understand this.

MechCoder replied:
It's been more than a year but the rationale was something like this.

Liblinear does not solve for the true multinomial loss and does a OvR, it now raises an error in sklearn if the loss is multinomial and the solver is liblinear.

In a OvR approach, for the other solvers if the class_weights are provided in a dict format, the class_weight are converted to sample_weights and then used. But liblinear does not support providing sample_weights, right now (https://github.com/scikit-learn/scikit-learn/pull/5274). However, in the future after the PR gets merged, we can do away with this.

amueller replied:
But it does support ``class_weight``, right? It's odd to support ``balanced`` but not dicts.

TomDLT replied:
here is the problem : 
- `_fit_liblinear(loss='logistic_regression')` does not handle `sample_weight`, but handles `class_weight` directly, and only with binary problems (!) .

- SAG, LBFGS and Newton-cg solvers handles `class_weight` by putting the weights into the `sample_weight`.

That is why we have a different handling of `class_weight` for liblinear.
___
Then, there are 3 cases, with `solver='liblinear'`:
 1. **multinomial case**, but it raise an error in `_check_solver_option`.
 2. **binary case**, then the `class_weight` dictionnary is reconstructed with keys -1 and 1, probably in order to match the API of `_fit_liblinear`
 3. **OvR with `n_classes > 2` case**, and then it raises an error (@MechCoder). The reason is that the other solvers use the class weight for each sample, but it is not possible in `fit_liblinear` since it does not handles `sample_weight` yet. With the 'balanced' case, the class weights are computed **after** the binarization one-vs-rest, which can be used in the same way by all solvers, liblinear included.

___
**Conclusion:** We probably just need to merge #5274 to be consistent with all solvers in `LogisticRegression`

The inconsistency for `class_weight='balanced'` in the OvR case (inconsistency between SGDClassifier, LinearSVC and LogisticRegression) is another topic (see https://github.com/scikit-learn/scikit-learn/pull/5008#issuecomment-139288413)

agramfort replied:
what's left to do for https://github.com/scikit-learn/scikit-learn/pull/5274
?

TomDLT replied:
I think #5274 is ready, but it only adds samples weights in `LogisticRegression` and not in other classes that use liblinear.

agramfort replied:
one thing at a time right?

TomDLT replied:
Sure

MechCoder replied:
I was about to summarize it, but thanks @TomDLT for doing it.. that was the issue. Once 5724 is merged, we can get rid of this check.

MechCoder replied:
Whether or not is the best way to do things is of course a different issue.

TomDLT commented:
I added the test from #5420, which tests the bug detailed in #5415.
I also changed the variable names to distinguish `y_bin` and `Y_bin`
____

About the discussion in #5420:
-This PR tests that these two classifiers are equivalent:
```python
class_weights = compute_class_weight("balanced", np.unique(y), y)
clf1 = LogisticRegression(multi_class="multinomial", class_weight="balanced")
clf2 = LogisticRegression(multi_class="multinomial", class_weight=class_weights)
```
-These two classifiers are still **NOT** equivalent:
```python
class_weights = compute_class_weight("balanced", np.unique(y), y)
clf1 = LogisticRegressionCV(class_weight="balanced")
clf2 = LogisticRegressionCV(class_weight=class_weights)
```
Indeed, the first one computes different `class_weights` for every folds, whereas the second one uses the same `class_weights` for every folds. That is why we can't remove [this line](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/tests/test_common.py#L116)

kastnerkyle commented on sklearn/linear_model/logistic.py:
```diff
@@ -623,20 +623,20 @@ def logistic_regression_path(X, y, pos_class=None, Cs=10, fit_intercept=True,
         y_bin[~mask] = -1.
+        # for compute_class_weight
+
+        if class_weight in ("auto", "balanced"):
```
Is there any internal comment convention for things that are deprecated? ```# TODO:``` looks horrible but some kind of tag could be nice.

amueller replied:
Comment with the version is most helpful. ``#TODO`` is fine. I grep for the version so ``0.19`` in this case I guess

TomDLT replied:
I'll add a comment

kastnerkyle commented:
Besides my comment this seems pretty solid.

amueller commented:
Can you remove the line in common tests that excludes ``LogisticRegressionCV`` from the ``class_weight`` tests? That should work now, right?

TomDLT commented:
> Can you remove the line in common tests that excludes LogisticRegressionCV from the class_weight tests? That should work now, right?

Are you talking about [this line](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/tests/test_common.py#L116)? It has been removed.

amueller commented:
Yes. Sorry I overlooked that.

amueller commented on sklearn/linear_model/logistic.py:
```diff
@@ -610,7 +610,7 @@ def logistic_regression_path(X, y, pos_class=None, Cs=10, fit_intercept=True,
                                  "class_weight of type dict. Use the lbfgs, "
                                  "newton-cg or sag solvers or set "
-                                 "class_weight='auto'")
+                                 "class_weight='balanced'")
```
Sorry if I'm being stupid, but I still don't understand why this is. Why can ``liblinear`` do "balanced" but not a dict? Maybe this is outside the scope of this fix, but I feel the logic here is hard to follow. Some of the class weight is handled here, and other parts are handled further down in the ``if mulitclass == 'ovr'``

TomDLT replied:
This is not clear indeed, I am trying to understand it

TomDLT replied:
I think this is related to the different handling of handling `class_weight` with `multi_class='ovr'`

amueller replied:
It's kind of unrelated to what you are fixing, I'm not sure if we should address it in this PR or not. It just seems odd.

TomDLT replied:
cf https://github.com/scikit-learn/scikit-learn/pull/5008#discussion_r42640364

Indeed, this is not related to this PR

amueller commented on sklearn/linear_model/logistic.py:
```diff
@@ -1538,6 +1538,15 @@ def fit(self, X, y, sample_weight=None):
                              "dict or 'balanced'")
 
+        # compute the class weights for the entire dataset y
+        if self.class_weight in ("auto", "balanced"):
```
why do we need this here?  logistic_regression_path later on should do that, right?

TomDLT replied:
Yes, but in ` logistic_regression_path` it would be computed on the folds, and not on the entire dataset, which is the cause of the "bug" #5450, and the cause of the line in common tests that excludes LogisticRegressionCV from the class_weight tests.

amueller replied:
I'm not sure that it is a bug. But I don't have a strong opinion. We could also make the ``class_weight`` pass by making the list of ``Cs`` just one long. Hm maybe the current master behavior is the same as the ``class_weight="balanced_bootstrap"`` in random forests? If so then we should do it as you propose here.

amueller replied:
well I guess it is a slightly different case.... hum....

amueller commented:
ok LGTM

agramfort commented:
+1 for merge

amueller commented:
backporting

amueller commented:
@TomDLT can you please add a whatsnew for 0.17?

ogrisel commented:
Backported by @amueller to 0.17.X as 289c0a3bb759d8685fd38aefdd8180f44eb665df.
